### PR TITLE
GuiIDProperty and ContainerProvider Impl.

### PR DESCRIPTION
**Pull request by Faithcaio**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1502) | *Common*

### CODE REVIEW

File: src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java [hunk 1]
```diff
@@ -138,36 +141,12 @@ public Container createContainer(InventoryPlayer playerInventory, EntityPlayer p
...
-        } else if (this.archetype == InventoryArchetypes.BEACON) {
-            return "minecraft:beacon";
-        } else if (this.archetype == InventoryArchetypes.HORSE || this.archetype == InventoryArchetypes.HORSE_WITH_CHEST) {
-            return "EntityHorse";
-        } else if (this.archetype == InventoryArchetypes.VILLAGER) {
-            return "minecraft:villager";
-        } else {
-            return "minecraft:chest";
-        }
+        return archetype.getProperty(GuiIDProperty.class, GuiIDProperty.class.getSimpleName().toLowerCase()).map(p -> p.getValue()).orElse("minecraft:chest");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 2]
```diff
@@ -84,4 +96,13 @@ public String getName() {
     public Map<String, InventoryProperty<String, ?>> getProperties() {
         return this.properties;
     }
+
+    @Override
+    public Container getContainer(Inventory inventory, Player player) {
+        if (containerProvider == null)
+        {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: Brace is on the wrong line.

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 3]
```diff
@@ -26,25 +26,37 @@
...
 
     private final String id;
     private final String name;
     private final List<InventoryArchetype> types;
     private final Map<String, InventoryProperty<String, ?>> properties;
+    private BiFunction<Inventory, Player, Container> containerProvider;
 
-    public CompositeInventoryArchetype(String id, String name, List<InventoryArchetype> types, Map<String, InventoryProperty<String, ?>> properties) {
+    public CompositeInventoryArchetype(String id, String name, List<InventoryArchetype> types, Map<String, InventoryProperty<String, ?>> properties,
+            BiFunction<Inventory, Player, Container> containerProvider) {
```
Comment (liach) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: `@Nullable`

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 4]
```diff
@@ -84,4 +96,12 @@ public String getName() {
     public Map<String, InventoryProperty<String, ?>> getProperties() {
         return this.properties;
     }
+
+    @Override
+    public Container getContainer(Inventory inventory, Player player) {
+        if (containerProvider == null) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 5]
```diff
@@ -84,4 +96,12 @@ public String getName() {
...
+            return ((Container) new CustomContainer(((EntityPlayer) player), ((CustomInventory) inventory)));
+        }
+        return containerProvider.apply(inventory, player);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/archetype/SpongeInventoryArchetypeBuilder.java [hunk 6]
```diff
@@ -64,7 +70,7 @@
     public InventoryArchetype build(String id, String name) {
         // TODO register archetype
         // TODO events
-        return new CompositeInventoryArchetype(id, name, types, properties);
+        return new CompositeInventoryArchetype(id, name, types, properties, containerProvider);
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/SpongeInventoryArchetypeBuilder.java]: `this.`

File: src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java [hunk 7]
```diff
@@ -138,36 +141,13 @@ public Container createContainer(InventoryPlayer playerInventory, EntityPlayer p
...
-            return "minecraft:beacon";
-        } else if (this.archetype == InventoryArchetypes.HORSE || this.archetype == InventoryArchetypes.HORSE_WITH_CHEST) {
-            return "EntityHorse";
-        } else if (this.archetype == InventoryArchetypes.VILLAGER) {
-            return "minecraft:villager";
-        } else {
-            return "minecraft:chest";
-        }
+        return this.archetype.getProperty(GuiIDProperty.class, GuiIDProperty.class.getSimpleName().toLowerCase()).map(p -> p.getValue()).orElse
+                ("minecraft:chest");
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java]: This formatting is weird - you should move the entire call (`.orElse("minecraft:chest");` to this line, or move this to the previous line.

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java (hunk 7 above)
Comment (Faithcaio) [reply src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java]: damn autoformatting

File: src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java [hunk 8]
```diff
@@ -139,36 +142,19 @@ public Container createContainer(InventoryPlayer playerInventory, EntityPlayer p
...
-        } else if (this.archetype == InventoryArchetypes.BEACON) {
-            return "minecraft:beacon";
-        } else if (this.archetype == InventoryArchetypes.HORSE || this.archetype == InventoryArchetypes.HORSE_WITH_CHEST) {
-            return "EntityHorse";
-        } else if (this.archetype == InventoryArchetypes.VILLAGER) {
-            return "minecraft:villager";
-        } else {
-            return "minecraft:chest";
-        }
+        return this.archetype.getProperty(GuiIdProperty.class, GuiIdProperty.class.getSimpleName().toLowerCase()).map(p -> p.getValue()).orElse
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/custom/CustomInventory.java]: Worth caching the value of `GuiIdProperty.class.getSimpleName().toLowerCase()`?

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainerHorseChest.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+
+import java.util.Optional;
+
+@Mixin(ContainerHorseChest.class)
+public abstract class MixinContainerHorseChest implements CarriedInventory {
+
+    @Override
+    public Optional getCarrier() {
+        return Optional.empty(); // TODO can we get the horse in here somehow?
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainerHorseChest.java]: Could, technically, throw a mixin into `net.minecraft.entity.passive.AbstractHorse#initHorseChest` to set the carrier?

File: src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainerHorseChest.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package org.spongepowered.common.mixin.core.item.inventory;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/mixin/core/item/inventory/MixinContainerHorseChest.java:25]: ~~Missing license header.~~ :white_check_mark:

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 11]
```diff
@@ -26,25 +26,32 @@
...
 
+import javax.annotation.Nullable;
+
 public class CompositeInventoryArchetype implements InventoryArchetype {
 
     private final String id;
     private final String name;
     private final List<InventoryArchetype> types;
     private final Map<String, InventoryProperty<String, ?>> properties;
+    private ContainerProvider containerProvider;
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: `@Nullable`

File: src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java [hunk 12]
```diff
@@ -85,4 +92,16 @@ public String getName() {
     public Map<String, InventoryProperty<String, ?>> getProperties() {
         return this.properties;
     }
+
+    public ContainerProvider getContainerProvider() {
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/item/inventory/archetype/CompositeInventoryArchetype.java]: `@Nullable`

File: src/main/java/org/spongepowered/common/registry/type/item/GuiIdRegistryModule.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        this.guiIdMap.put(guiId.getId(), guiId);
+    }
+
+    private GuiIdRegistryModule() {
+    }
+
+    private static final class Holder {
+        static final GuiIdRegistryModule INSTANCE = new GuiIdRegistryModule();
+    }
+}
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/registry/type/item/GuiIdRegistryModule.java]: newline

**Review by kashike — CHANGES_REQUESTED** [review]

**Pull request merged**